openstf/pipeline/optimize_hyperparameters.py

```diff
@@ -0,0 +1,91 @@
+from openstf.model.objective_creator import ObjectiveCreator
+import pandas as pd
+import optuna
+import structlog
+
+from openstf.feature_engineering.feature_applicator import TrainFeatureApplicator
+from openstf.validation import validation
+
+# This is required to disable the default optuna logger and pass the logs to our own
+# structlog logger
+optuna.logging.enable_propagation()         # Propagate logs to the root logger.
+optuna.logging.disable_default_handler()    # Stop showing logs in sys.stderr.
+
+logger = structlog.get_logger(__name__)
+
+# See https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.optimize
+N_TRIALS: int = 8   # The number of trials.
+TIMEOUT: int = 200  # Stop study after the given number of second(s).
+
+def optimize_hyperparameters_pipeline(
+    pj: dict,
+    input_data: pd.DataFrame,
+    n_trials: int = N_TRIALS,
+    timeout: int = TIMEOUT
+):
+    """Optimize hyperparameters pipeline.
+
+    Args:
+        pj (dict): Prediction job
+        input_data (pd.DataFrame): Raw training input data
+        n_trials (int, optional): The number of trials. Defaults to N_TRIALS.
+        timeout (int, optional): Stop study after the given number of second(s).
+            Defaults to TIMEOUT.
+
+    Raises:
+        ValueError: If the input_date is insufficient.
+
+    Returns:
+        dict: Optimized hyperparameters.
+    """
+    # Validate and clean data
+    validated_data = validation.clean(validation.validate(input_data))
+
+    # Check if sufficient data is left after cleaning
+    if not validation.is_data_sufficient(validated_data):
+        raise ValueError(
+            f"Input data is insufficient for {pj['name']} "
+            f"after validation and cleaning"
+        )
+
+    input_data_with_features = TrainFeatureApplicator(horizons=[0.25, 24]).add_features(
+        input_data
+    )
+
+    # Create relevant model
+    Objective = ObjectiveCreator.create_objective(model_type=pj["model"])
```

Comment:
Objective should be objective (this also generates a code smell)

Reply:
hmm..okay. Since Objective is a class and not an instance of a class to me it makes sense to use a capital here. But I can also call it `objective_constructor` or `objective_class` to prevent the smell.

Reply:
I used `objective` and added a comment that this is a class and not an instance to prevent confusion.